File: src/main/java/org/usfirst/irs1318/gamesim/engine/actions/Pick.java
package org.usfirst.irs1318.gamesim.engine.actions;

import org.usfirst.irs1318.gamesim.engine.Action;
import org.usfirst.irs1318.gamesim.engine.Containable;
import org.usfirst.irs1318.gamesim.engine.Container;

import java.util.Optional;

/**
 * The actor retrieves a containers from the origin container and stores it in its own container.
 */
public class Pick extends Action {
    public String payloadType;
    public Container source;
    public Container failure;
    public int quantity;

    @Override
    public void performSuccessAction() {
        for (int i = 0; i < quantity; i++) {
            Optional<Containable> payload = source.retrieveContents();
            payload.ifPresent(containable -> actor.addPayload(payloadType, containable));
        }
    }

    @Override
    public void performFailureAction() {
        for (int i = 0; i < quantity; i++) {
            source.retrieveContents().ifPresent(containable -> failure.placeContent(containable));
        }
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/action/Pick.scala
package org.usfirst.irs1318.gamesim.engine.action

import org.usfirst.irs1318.gamesim.engine.GameEngine
import org.usfirst.irs1318.gamesim.engine.actors.Actor
import org.usfirst.irs1318.gamesim.engine.container.Container
import org.usfirst.irs1318.gamesim.engine.objective.TaskDetails

case class Pick(source: Container, failure: Container, actor: Actor, taskDetails: TaskDetails) extends Action {
  override def run(state: GameEngine.State): GameEngine.State = ???
}
